src/python/grpcio/grpc/_cython/_cygrpc/aio/grpc_aio.pyx.pxi

```diff
@@ -13,25 +13,97 @@
+
+    # Picks the engine for gRPC AsyncIO Stack
+    for engine_type in AsyncIOEngine:
+        if engine_type.value == _GRPC_ASYNCIO_ENGINE:
+            grpc_aio_engine = engine_type
+            break
+    if grpc_aio_engine is None or grpc_aio_engine is AsyncIOEngine.DEFAULT:
+        grpc_aio_engine = AsyncIOEngine.CUSTOM_IO_MANAGER
+
+    # Anchors the event loop that the gRPC library going to use.
+    _grpc_aio_loop = asyncio.get_event_loop()
+    _event_loop_thread_ident = threading.current_thread().ident
+
+    if grpc_aio_engine is AsyncIOEngine.CUSTOM_IO_MANAGER:
+        # Activates asyncio IO manager.
+        # NOTE(lidiz) Custom IO manager must be activated before the first
+        # `grpc_init()`. Otherwise, some special configurations in Core won't
+        # pick up the change, and resulted in SEGFAULT or ABORT.
+        install_asyncio_iomgr()
+
+        # TODO(https://github.com/grpc/grpc/issues/22244) we need a the
+        # grpc_shutdown_blocking() counterpart for this call. Otherwise, the gRPC
+        # library won't shutdown cleanly.
+        grpc_init()
+
+        # Timers are triggered by the Asyncio loop. We disable
+        # the background thread that is being used by the native
+        # gRPC iomgr.
+        grpc_timer_manager_set_threading(False)
+
+        # gRPC callbaks are executed within the same thread used by the Asyncio
+        # event loop, as it is being done by the other Asyncio callbacks.
+        Executor.SetThreadingAll(False)
+    else:
+        # TODO(https://github.com/grpc/grpc/issues/22244) we need a the
+        # grpc_shutdown_blocking() counterpart for this call. Otherwise, the gRPC
+        # library won't shutdown cleanly.
+        grpc_init()
+
+    _grpc_aio_initialized = False
+
+
+def grpc_aio_loop():
+    """Returns the one-and-only gRPC Aio event loop."""
+    return _grpc_aio_loop
+
 
-    install_asyncio_iomgr()
-    grpc_init()
+def grpc_schedule_coroutine(object coro):
+    """Thread-safely schedules coroutine to gRPC Aio event loop.
 
-    # Timers are triggered by the Asyncio loop. We disable
-    # the background thread that is being used by the native
-    # gRPC iomgr.
-    grpc_timer_manager_set_threading(False)
+    If invoked within the same thread as the event loop, return an
+    Asyncio.Task. Otherwise, return a concurrent.futures.Future (the sync
+    Future). For non-asyncio threads, sync Future objects are probably easier
+    to handle (without worrying other thread-safety stuff).
+    """
+    if _event_loop_thread_ident != threading.current_thread().ident:
+        return asyncio.run_coroutine_threadsafe(coro, _grpc_aio_loop)
+    else:
+        return _grpc_aio_loop.create_task(coro)
 
-    # gRPC callbaks are executed within the same thread used by the Asyncio
-    # event loop, as it is being done by the other Asyncio callbacks.
-    Executor.SetThreadingAll(False)
 
-    _grpc_aio_initialized = 1
+def grpc_call_soon_threadsafe(object func, *args):
```

Comment:
why not `cdef`?

Reply:
Because `*args` doesn't work with `cdef`.